evdns.c

```diff
@@ -2173,6 +2183,84 @@ evdns_request_timeout_callback(evutil_socket_t fd, short events, void *arg) {
 	}
 	EVDNS_UNLOCK(base);
 }
+static void evdns_request_read_cb_tcp(struct bufferevent *bev, void *ctx){
+    struct request *req = (struct request *)ctx;
+    struct evbuffer *buf;
+    size_t size;
+    uint16_t in_len; // in tcp DNS packet is prefixed with 2 byte length
+    unsigned char *pp;
+
+    if( !req )
+        goto err;
+    buf = bufferevent_get_input(bev);
+    if( !buf )
+        goto err;
+
+    size = evbuffer_get_length(buf);
+    if( size < sizeof(in_len) ){ // not enough data received
+        return;
+    }
+    if( evbuffer_copyout(buf, &in_len, sizeof(in_len)) != sizeof(in_len) )
+        goto err;
+    in_len = ntohs(in_len);
+    if( size < in_len + sizeof(in_len)){ // not enough data received
+        return;
+    }
+    if( evbuffer_drain(buf, sizeof(in_len)) )
+        goto err;
+    pp = evbuffer_pullup(buf, in_len);
+    if( !pp )
+        goto err;
+    reply_parse(req->base, pp, in_len);
+err:
+    if( bev )
+        bufferevent_free(bev);
+    return;
+}
+static void evdns_request_event_cb_tcp(struct bufferevent *bev, short what, void *ctx){
+    struct request *req = (struct request *)ctx;
+    if (what & (BEV_EVENT_ERROR | BEV_EVENT_READING | BEV_EVENT_WRITING | BEV_EVENT_TIMEOUT )) {
+        bufferevent_free(bev);
+        if( req ) {
+    	    reply_schedule_callback(req, 0, DNS_ERR_TRUNCATED, NULL);
+       		request_finished(req, &REQ_HEAD(req->base, req->trans_id), 1);  
+        }
+    }
+}
+
+/* try to send a request to a given server over tcp. */
+/* */
+/* return: */
+/*   0 ok */
+/*   1 temporary failure */
+/*   2 other failure */
+static int
+evdns_request_transmit_to_tcp(struct request *req, struct nameserver *server) {
+    struct bufferevent *be;
+    uint16_t len;
+    ASSERT_LOCKED(req->base);
+    ASSERT_VALID_REQUEST(req);
+    be = bufferevent_socket_new(req->base->event_base, -1,  BEV_OPT_CLOSE_ON_FREE);
```

Comment:
it is too huge overhead creating tcp for each request, this should be done once (with some number of retries)